### PR TITLE
legendre polynomials and gaussian quadrature

### PR DESCRIPTION
**Pull request by hsnyder**
- Added preliminary implementation for Legendre polynomials
- Added preliminary implementation for Legendre polynomial first derivatives
- Added preliminary implementation for Gauss-Legendre quadrature points/weights
- Added preliminary implementation for Gauss-Legendre-Lobatto quadrature points/weights

link to proposal: https://github.com/fortran-lang/stdlib/issues/277

### CODE REVIEW

File: src/stdlib_quadrature_gauss.f90
Comment (ivan-pi) [inline src/stdlib_quadrature_gauss.f90:111]: I think here it should be `0.5_dp` to avoid mixed precision arithmetic. It might be worth checking that `x(1) == interval(1)` and `x(size(x)) == interval(2)`. I imagine this might not be the case due to rounding errors.

File: src/stdlib_quadrature_gauss.f90
Comment (ivan-pi) [inline src/stdlib_quadrature_gauss.f90:46]: `x(N/2+1) = 0` (implicit type casting)

File: src/stdlib_quadrature_gauss.f90
Comment (ivan-pi) [inline src/stdlib_quadrature_gauss.f90:83]: A comment containing the source for the initial guess could be helpful.

File: src/stdlib_quadrature_gauss.f90
Comment (ivan-pi) [inline src/stdlib_quadrature_gauss.f90:20]: I think this should be
```
x(1) = 0
w(1) = 2
```

File: src/stdlib_quadrature_gauss.f90
Comment (ivan-pi) [inline src/stdlib_quadrature_gauss.f90:23]:
```
x(1) = -sqrt(1.0_dp/3.0_dp)
x(2) = -x(1)
w = 1
```
Perhaps this is not worth bothering about. Hopefully, the compiler does not generate temporaries.

**Comment by certik** [comment]
I have a comment regarding the implementation --- what is the accuracy? I had bad experience with higher order (say `p=64`) that wasn't accurate to machine accuracy with such methods. So we resorted to calculate the weights and points using high precision and generate code that hard codes the numbers valid to all digits. But this comment is an implementation, which is minor.

Let's mainly discuss the API first. I would remove `interval` as doing `[-1, 1]` seems like a default?

**Comment by ivan-pi** [comment]
> Let's mainly discuss the API first. I would remove `interval` as doing `[-1, 1]` seems like a default?

The `interval` is an optional argument: 
```
pure module subroutine gauss_legendre_fp64 (x, w, interval)
        real(dp), intent(out) :: x(:), w(:)
        real(dp), intent(in), optional :: interval(2)
```

The default is `[-1,1]`. At least in (chemical) engineering problems, shifted Legendre polynomials are sometimes used. I find the `interval` argument handy for such cases. Of course one could argue, that the user of the quadrature should be responsible for such things.

I am afraid the `intent(out)` attribute for `x`, and `w` is wrong. If these are allocatable arrays in the calling program, then `gauss_legendre` will not work as intended. Adding an integer variable, and using fixed size arrays would be the most flexible solution.

**Comment by nncarlson** [comment]
> I am afraid the `intent(out)` attribute for `x`, and `w` is wrong. If these are allocatable arrays in the calling program, then `gauss_legendre` will not work as intended.

How so?  I think you are mistaken here. Whether the actual arguments are allocatable is irrelevant. Are you thinking that the `intent(out)` causes them to be deallocated?

**Comment by ivan-pi** [comment]
@nncarlson, you are right.

**Comment by milancurcic** [comment]
Thank you @hsnyder. I didn't do a proper review yet, but two things stood out on a cursory skim through:

1. stdlib_functions.f90 for the source file and `stdlib_functions` for the module are not informative IMO. Is there some descriptive name we can use? What kind of functions are these?
2. Please use [lowercase for all variables](https://github.com/fortran-lang/stdlib/blob/master/STYLE_GUIDE.md#variable-and-procedure-naming). I think uppercase is okay for constants (`parameter`) although we don't have an established position on this yet.

**Comment by ivan-pi** [comment]
>   2. Please use [lowercase for all variables](https://github.com/fortran-lang/stdlib/blob/master/STYLE_GUIDE.md#variable-and-procedure-naming). I think uppercase is okay for constants (`parameter`) although we don't have an established position on this yet.

Unfortunately, the lower-case `l` is sometimes frowned upon due to its potential for mix up with `1`. Perhaps `p` (for the polynomial) and `dp` (for its derivative) can be used as substitutes.

**Comment by awvwgk** [comment]
> Perhaps `p` (for the polynomial) and `dp` (for its derivative) can be used as substitutes.

`dp` will unfortunately clash with the kind parameter for double precision reals. More than one or two letters might be advisable instead.

**Comment by ivan-pi** [comment]
> `dp` will unfortunately clash with the kind parameter for double precision reals. More than one or two letters might be advisable instead.

Right... My next suggestion then are `pn` and `dpn`, which are unsurprisingly also used in the code by [Stroud & Secrest](https://github.com/ivan-pi/stroud_quad/blob/b36291d94ffa8a81e5e8aed6d7afb37bc4c0b0e7/legacy/recur.f#L1).

**Comment by hsnyder** [comment]
Thanks everyone for the input so far!

> Thank you @hsnyder. I didn't do a proper review yet, but two things stood out on a cursory skim through:
> 
>     1. stdlib_functions.f90 for the source file and `stdlib_functions` for the module are not informative IMO. Is there some descriptive name we can use? What kind of functions are these?

The only thing I can really think of is "special functions". Acceptable?

**Comment by arjenmarkus** [comment]
"Special functions" is the general term that is used for this part of
mathematics. So I'd say it is certainly acceptable.

Op ma 8 feb. 2021 om 21:04 schreef Harris Snyder <notifications@github.com>:

**Review by arjenmarkus — COMMENTED** [review]
Just a remark: the do-loop at line 33 runs from 0 to newton_iters-1. As the do-loop variable is not explicitly used, I would say: let it run from 1 to newton_iters instead.

**Comment by hsnyder** [comment]
> Just a remark: the do-loop at line 33 runs from 0 to newton_iters-1. As the do-loop variable is not explicitly used, I would say: let it run from 1 to newton_iters instead.

haha you can tell I'm a C programmer... I will fix this

**Comment by hsnyder** [comment]
sorry for the delay I have been very busy. other than resolving the makefile conflicts, what are the next steps towards getting this merged? I will try to get to the conflicts in the makefiles sometime this week but if I can't and someone else wants to, please feel free :)

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:2]: Remove this line to solve the compilation error with GCC 10.2
```suggestion
```

File: src/stdlib_specialfunctions_legendre.f90
Comment (awvwgk) [inline src/stdlib_specialfunctions_legendre.f90:31]: I get four warnings on implicit conversion here:
```
../src/stdlib_specialfunctions_legendre.f90:31:31:

   31 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                               1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_specialfunctions_legendre.f90:31:51:

   31 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                                                   1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_specialfunctions_legendre.f90:31:55:

   31 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                                                       1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_specialfunctions_legendre.f90:31:71:

   31 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                                                                       1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```
Can we make those integer real conversion explicit?

File: src/stdlib_specialfunctions_legendre.f90
Comment (awvwgk) [inline src/stdlib_specialfunctions_legendre.f90:32]:
```
../src/stdlib_specialfunctions_legendre.f90:32:45:

   32 |                         dleg = dleg_down2 + (2*i-1)*leg_down1
      |                                             1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```
Can we make the integer real conversion explicit?

File: src/stdlib_specialfunctions_legendre.f90
Comment (awvwgk) [inline src/stdlib_specialfunctions_legendre.f90:62]: Another bunch of conversion warnings, can we make those explicit?
```
../src/stdlib_specialfunctions_legendre.f90:62:31:

   62 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                               1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_specialfunctions_legendre.f90:62:51:

   62 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                                                   1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_specialfunctions_legendre.f90:62:55:

   62 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                                                       1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_specialfunctions_legendre.f90:62:71:

   62 |                         leg = (2*i-1)*x*leg_down1/i - (i-1)*leg_down2/i
      |                                                                       1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:88]: More implicit conversion warnings here:
```
../src/stdlib_quadrature_gauss.f90:88:36:

   88 |                     x(i+1) = -cos( (i+0.25_dp)*pi/n  - 3/(8*n*pi*(i+0.25_dp)))
      |                                    1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:88:66:

   88 |                     x(i+1) = -cos( (i+0.25_dp)*pi/n  - 3/(8*n*pi*(i+0.25_dp)))
      |                                                                  1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:88:60:

   88 |                     x(i+1) = -cos( (i+0.25_dp)*pi/n  - 3/(8*n*pi*(i+0.25_dp)))
      |                                                            1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:88:51:

   88 |                     x(i+1) = -cos( (i+0.25_dp)*pi/n  - 3/(8*n*pi*(i+0.25_dp)))
      |                                                   1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:83]: This doesn't look too problematic, but we should minimize the compiler warnings generated by stdlib as much as possible:
```
../src/stdlib_quadrature_gauss.f90:83:36:

   83 |                 w(1)   =  2._dp/(n*(n+1._dp))
      |                                    1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:83:34:

   83 |                 w(1)   =  2._dp/(n*(n+1._dp))
      |                                  1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:84]: Again, not too problematic, but better explicit than implicit:
```
../src/stdlib_quadrature_gauss.f90:84:36:

   84 |                 w(n+1) =  2._dp/(n*(n+1._dp))
      |                                    1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:84:34:

   84 |                 w(n+1) =  2._dp/(n*(n+1._dp))
      |                                  1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:99]: More implicit conversion here
```
../src/stdlib_quadrature_gauss.f90:99:41:

   99 |                     w(i+1)   = 2._dp/(n*(n+1._dp)*leg**2)
      |                                         1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:99:39:

   99 |                     w(i+1)   = 2._dp/(n*(n+1._dp)*leg**2)
      |                                       1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:107]:
```
../src/stdlib_quadrature_gauss.f90:107:43:

  107 |                     w(n/2+1)   = 2._dp/(n*(n+1._dp)*leg**2)
      |                                           1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:107:41:

  107 |                     w(n/2+1)   = 2._dp/(n*(n+1._dp)*leg**2)
      |                                         1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

File: src/stdlib_quadrature_gauss.f90
Comment (awvwgk) [inline src/stdlib_quadrature_gauss.f90:31]:
```
../src/stdlib_quadrature_gauss.f90:31:50:

   31 |                     x(i+1) = -cos((2*i+1)/(2._dp*n+2._dp) * pi)
      |                                                  1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
../src/stdlib_quadrature_gauss.f90:31:35:

   31 |                     x(i+1) = -cos((2*i+1)/(2._dp*n+2._dp) * pi)
      |                                   1
Warning: Conversion from INTEGER(4) to REAL(8) at (1) [-Wconversion-extra]
```

**Review by awvwgk — COMMENTED** [review]
GCC flags a lot of implicit conversions here, they are probably all okay and not harmful. Still, can we make those explicit to reduce the noise to signal ratio of compiler warnings generated by stdlib?

**Comment by awvwgk** [comment]
Most of the GCC warnings are generated because we are always enforcing it from our build files with
https://github.com/fortran-lang/stdlib/blob/0886501d057656cedf1806ebcd1ba2c2afc21e6a/CMakeLists.txt#L24

We get 23 possibly “false positive” warnings for this patch from `-Wconversion-extra` with GCC. There are only two ways forward IMO, either we did enable this warning because we want to avoid implicit conversion, than this patch has to account for this and make all conversions explicit, or we regard those warnings as not useful because they are false positives and delete this option from build files.

**Comment by milancurcic** [comment]
I don't see a problem with these warnings. We could have the flag only in development and remove it for release build. I personally prefer implicit type casts where safe as it allows for cleaner code. But I'm not opposed to explicit conversion if others prefer it. Something to think about to add to the style guide.

**Comment by ivan-pi** [comment]
In this usage case I'm in favor of implicit casts because they make sure the expression match closely the original mathematical form of the [recurrence relations](https://en.wikipedia.org/wiki/Legendre_polynomials#Recurrence_relations).

**Comment by awvwgk** [comment]
> We could have the flag only in development and remove it for release build. I personally prefer implicit type casts where safe as it allows for cleaner code

This is actually not true, the `-Wconversion-extra` is added unconditionally for all builds. Therefore, I assumed it should always be honored, thanks for clarifying that this is not the case. I'll open a new issue to discuss our compilation profiles than.

**Review by awvwgk — APPROVED** [review]
With my confusion about the strictness of warnings resolved, this patch looks good to be merged.

**Comment by certik** [comment]
I think this looks good, but we need to add tests, showing how to use it and also ensuring it actually works as expected for some rules.

**Comment by hsnyder** [comment]
awvwgk, thanks very much for fixing those build issues. @certik I can add some tests and documentation, though probably not for another week or so. if you have any specific tests you'd like me to add please let me know.

**Comment by certik** [comment]
I don't have any specific tests, just any tests that would run this say for order 3 and 5 and 64 and check the correct result/accuracy.

**Comment by awvwgk** [comment]
What is the current status of this patch? @hsnyder can we do anything to help you move this pull request forward?

**Comment by hsnyder** [comment]
@awvwgk sorry for the delay. I think test cases and a spec are the outstanding items. In principle anyone could write them but it would obviously be easiest for me to do so. I’ve been very busy for the past few weeks but things are starting to ease up. I think I can finish this off this week. I will post again when I’ve done so.

**Comment by hsnyder** [comment]
Sorry for the delay! I've added some tests and specs for the new quadrature functions and the special functions module. Please let me know if any revisions are required.

*note*: I wasn't sure how to validate the the spec format is correct. I tried to follow the examples already present but I may have made errors and am not sure how to check it.

File: doc/specs/stdlib_quadrature.md
Comment (awvwgk) [inline doc/specs/stdlib_quadrature.md:229]: This should be preferably a standalone demo program

File: doc/specs/stdlib_quadrature.md
Comment (awvwgk) [inline doc/specs/stdlib_quadrature.md:250]:
```suggestion
`subroutine [[stdlib_quadrature(module):gauss_legendre_lobatto(interface)]] (x, w[, interval])`
```

File: doc/specs/stdlib_specialfunctions.md
Comment (awvwgk) [inline doc/specs/stdlib_specialfunctions.md:24]:
```suggestion
`result = [[stdlib_specialfunctions(module):legendre(interface)]] (n, x)`
```

File: doc/specs/stdlib_specialfunctions.md
Comment (awvwgk) [inline doc/specs/stdlib_specialfunctions.md:53]:
```suggestion
`result = [[stdlib_specialfunctions(module):dlegendre(interface)]] (n, x)`
```

**Review by awvwgk — APPROVED** [review]
Thanks a lot, looks good to me now. Only a few comments regarding the markup.

File: doc/specs/stdlib_quadrature.md
Comment (awvwgk) [inline doc/specs/stdlib_quadrature.md:227]:
```suggestion
	implicit none
```

File: doc/specs/stdlib_quadrature.md
Comment (awvwgk) [inline doc/specs/stdlib_quadrature.md:273]:
```suggestion
	implicit none
```

**Review by milancurcic — APPROVED** [review]
Thank you @hsnyder and reviewers!

I have one last minute question and suggestion: What do you think about renaming everything that's called `stdlib_specialfunctions*` to `stdlib_special*`? I find `specialfunctions` quite long and a bit hard to read because two words are together, and `special` is clear enough what it means IMO.

I think this PR can go forward either way, but I'd prefer if we called this just `stdlib_special`.

**Comment by hsnyder** [comment]
Hi @milancurcic , I prefer specialfunctions because of clarity, but I don't have strong feelings about it. On the one hand, I agree that "special" is shorter, easier to type and read, but on the other hand it's not very descriptive (I don't think "special" on it's own is very clear - the stdlib includes more than just math, so I think "special what?" is a perfectly valid reaction).

I'll wait ~ 24 hours for any additional opinions re: the name change and then I'll merge the request. Thanks very much for reviewing!

**Comment by arjenmarkus** [comment]
An alternative might be "stdlib_specfunc" - though perhaps that is too
reminiscent of other packages/libraries in other languages?

Op di 29 jun. 2021 om 00:31 schreef Harris Snyder ***@***.***

**Comment by hsnyder** [comment]
So it turns out I cannot merge this myself as I don't have write access. @milancurcic are you able to do so? Feel free to change the name if you want (though I still prefer "specialfunctions", or Arjen's suggestion instead of just "special"). 

Thanks!

**Comment by milancurcic** [comment]
Thanks, @hsnyder, I'll merge as is.

**Pull request merged**
